### ISSUE TITLE
Vim9: type of list based only on first item

### ISSUE THREAD

**Issue by lacygoill** [vim/vim#6650]
**Describe the bug**

In Vim9 script, the type of a list is based only on its first item.

**To Reproduce**

Run this shell command:

    vim -Nu NONE -S <(cat <<'EOF'
        vim9script
        def Func(numbers: list<float>)
        enddef
        Func([1.23, 'a', true])
    EOF
    )

No error is raised.

**Expected behavior**

`E1013` is raised:

    E1013: type mismatch, expected list<float> but got list<any>

**Environment**

 - Vim version: 8.2 Included patches: 1-1382
 - OS: Ubuntu 16.04.6 LTS
 - Terminal: XTerm(358)

**Additional context**

Here are other examples:
```vim
vim9script
def Func(numbers: list<float>)
enddef
Func(['a', 1.23, true])
```
    E1013: type mismatch, expected list<float> but got list<string>
                                                   ^--------------^
                                                          ✘
```vim
vim9script
def Func(numbers: list<float>)
enddef
Func([{}, 1.23, true])
```
    E1013: type mismatch, expected list<float> but got list<dict<unknown>>
                                                   ^---------------------^
                                                              ✘
I think the last 2 snippets should give this error message:

    E1013: type mismatch, expected list<float> but got list<any>
                                                   ^-----------^
                                                         ✔

**Comment by lacygoill** [comment]
I'm not sure this is a bug.  Maybe it would be too costly for Vim to check the type of every single item in a huge list; so it falls back on the first item and assume that all the remaining items share the same type.

If so, maybe this assumption should be documented; I don't think it is.

**Comment by brammool** [comment]
> I'm not sure this is a bug.  Maybe it would be too costly for Vim to
> check the type of every single element in a huge list; so it falls
> back on the first element and assume that all the remaining elements
> share the same type.
> 
> If so, maybe it should be documented; I don't think it is.

Currently the type of the first item is used.  But that's not
necessarily what we want.

At least, when defining a list of constants, the type of all items need
to be checked, this should give an error:

	let l: list<number> = [234, 'aasdf']

What if the type is not specified?  Without items:

	let l = []

We alread imply list<any>.  What if the types are mixed?

	let l = [234, 'aasdf']

Should we use the type from the first item and give an error for the
second?  That would avoid mistakes.  To allow for any item type:

	let l: list<any> = [234, 'aasdf']

This probably works best, using the implied type from the first item is
usually what we want and avoids having to add a type to get type
checking.

-- 
We are the Borg of GNU GPL.  We will assimilate your source code.
Resistance is futile.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

**Comment by lacygoill** [comment]
> At least, when defining a list of constants, the type of all items need
> to be checked, this should give an error:
>
>        let l: list<number> = [234, 'aasdf']

I agree, but to be consistent, an error should also be given here:
```vim
vim9script
def Func(l: list<number>)
enddef
Func([234, 'aasdf'])
```
---

> What if the types are mixed?
>
>         let l = [234, 'aasdf']
>
> Should we use the type from the first item and give an error for the
> second?  That would avoid mistakes.

I would have assumed that `list<any>` was inferred in this case, but being strict here may help find bugs earlier in the code.  I don't know what's best.

**Comment by brammool** [comment]
With a declaration we can specify the type and assign a list that starts with a different type:
        let var: list<any> = [234, 'asdf']
But when calling a function this is not possible:
        TheFunc([234, 'asdf'])
If we use the type of the first item, this will never work.  Also, if the first item is a variable without type:
        [g:somevar, 234]
Then the type is list<any>, but [234, g:somevar] again would not be possible.
Thus it looks like the type of the list must be determined by using the types of all items.

**Comment by brammool** [comment]
This was implemented in 8.2.1407
